### ISSUE TITLE
Adversarial Feature Matching for Text Generation

### ISSUE THREAD

**Issue by fhiyo** [fhiyo/ml_papers#52]
## 論文情報

<http://proceedings.mlr.press/v70/zhang17b.html>

Yizhe Zhang and Zhe Gan and Kai Fan and Zhi Chen and Ricardo Henao and Dinghan Shen and Lawrence Carin
Duke University, Durham, NC, 27708.

ICML 2017

## 一言で言うと

GANでテキストを生成する研究。GANの問題であるmode collapseとvanishing gradient descentの課題に取り組んだTextGANを提案している。

## 背景

BowmanらのCoNLL 2016 ([Generating sentences from a continuous space](https://www.aclweb.org/anthology/K16-1002/)) によれば、入力テキストから潜在空間への写像は多様体をなし、その範囲は空間全体から見て限られたものである。そのため潜在空間からのサンプリングによりテキストを生成しても、リアルな文が出力されない。同論文ではRNNベースのVAE-AEにより多様なテキストを生成しようと試みているが、隠れ変数の事後分布が適切に潜在空間に被覆していないという本質的な問題には対処できていない。

RNNによる生成にはexposure biasの問題があり、その問題を解決すべくBengioらのNIPS 2015 ([Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks](http://papers.nips.cc/paper/5956-scheduled-sampling-for-sequence-prediction-with-recurrent-neural-networks)) でscheduled samplingの手法が提案されている。しかしこの手法は安定しない。

GANはdiscriminatorがリアルデータと合成データの判別を行うように訓練されるので、exposure biasの問題を軽減されるような設定になっている。GANに強化学習の仕組みを取り入れたモデルも提案されているが、mode collapseとvanishing gradient descentの問題を抱えている。

本論文ではTextGANという上記のmode collapseとvanishing gradient descentの問題を軽減させるGANのモデルを提案している。

## 貢献

TextGANという上記のmode collapseとvanishing gradient descentの問題を軽減させるモデルを提案した点。
通常のGANが行う勾配計算に加えて、再生核ヒルベルト空間上でリアルデータと合成データを特徴量化したものの間で内積を取ったL\_{GAN}と、generator (G) の入力である潜在表現zとgeneratorから生成された文字列をCNNで特徴量化したtilde{f}を潜在空間に写像したhat{z}とのユークリッド距離を項に加えている。

![image](https://user-images.githubusercontent.com/13327262/72738472-fe1b3680-3be4-11ea-92e3-a8c55ffe02cb.png)

reconstruction lossとMMD lossはbinary classification lossの正則化項として働いており、2つのlossを項に入れることでfeatureは空間上により広く散らばっている様子を観測したらしい。

MMDを導入することでリアルデータと合成データの分布としての一致性を見ることができるので、mode collapseに陥りにくい。 (ref: [Generative moment matching networks (ICML 15)](https://arxiv.org/pdf/1502.02761.pdf))

vanishing gradient descentの問題をどのくらい改善している仕組みになっているかについてはあまり明示的に言及していない？

## 提案

![image](https://user-images.githubusercontent.com/13327262/72742073-543fa800-3bec-11ea-832d-5d2717ee8d9c.png)

提案モデルのD, Gは上図。

training techniqueとして、

- soft-argmax approximation

![image](https://user-images.githubusercontent.com/13327262/72743028-34a97f00-3bee-11ea-98be-28a9382c489a.png)

の形で出力tokenを決定する

- pre-training

CNN-LSTM autoencoderのタスクを事前に学習する

などを行っている。

## 実験・考察

BookCorpus datasetとArXiv datasetの2つを混ぜて訓練データとした (50万文ずつ使用)。

4つのlossでtextGANの性能を比較。

- Mean Matching (MM)
- Covariance Matching (CM)
- MMD
- MMD with compressed network (MMD-L)

また、baselineとしてAEモデル、VAE model、seqGANとも比較した。

評価指標はBLEUとKDE。

![image](https://user-images.githubusercontent.com/13327262/72745481-332e8580-3bf3-11ea-8678-f3a5399f28e1.png)

結果は上図。MMD-LがBLEU-2, 3, KDEで最もスコアが良い。

textGANの訓練後、Dの誤答は0.05しか無かった。これはGが局所解に落ちているからかもしれない。

![image](https://user-images.githubusercontent.com/13327262/72746733-2f503280-3bf6-11ea-9331-ffc176e79243.png)

text Aからtext Bまで、潜在空間上で直線に移動したときのサンプリング結果。textGANの方がAEよりも構文・意味ともにそれっぽい文になっているが、単語のチョイスや文構造が急に変化しているところもある。

## 補足・疑問など

table 3は、平滑性が場所によって異なっていることが言える？
